### PR TITLE
mgmt, support function app on ACA

### PR DESCRIPTION
**Pull request by XiaofeiCao**
# Description

address https://github.com/Azure/azure-sdk-for-java/issues/37156

This PR adds support for Function App hosted on ACA(Azure Container App) environment.
```java
FunctionApp functionApp = appServiceManager
    .functionApps()
    .define(webappName1)
    .withRegion(region)
    .withExistingResourceGroup(resourceGroup)
    // specifies the Azure Container Apps environment id
    .withManagedEnvironmentId(managedEnvironmentId)
    // scale settings
    .withMaxReplicas(10)
    .withMinReplicas(3)
    // docker image settings
    .withPublicDockerHubImage("mcr.microsoft.com/azure-functions/dotnet7-quickstart-demo:1.0")
    .create();
``` 
Above code corresponds to the [`http-trigger`](https://github.com/Azure/azure-functions-on-container-apps/tree/main#create-the-local-function-project-c-isolated---http-trigger) scenario.

`managedEnvironmentName` can't be modified once Function App's created.

docs: 
- [Function App on ACA managed environment](https://learn.microsoft.com/en-us/azure/azure-functions/functions-container-apps-hosting)
- [ACA managed environment](https://learn.microsoft.com/en-us/azure/container-apps/environment)
- [All hosting options for Function App](https://learn.microsoft.com/en-us/azure/azure-functions/functions-scale)

Please add an informative description that covers that changes made by the pull request and link all relevant issues.

If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java [hunk 1]
```diff
@@ -209,16 +275,92 @@ Mono<Indexable> submitAppSettings() {
...
+                    .createOrUpdateConfigurationWithResponseAsync(resourceGroupName(), name(), siteConfig),
+                SiteConfigResourceInner.class);
+        } else {
+            return super.createOrUpdateSiteConfig(siteConfig);
+        }
+    }
+
+    private <T> Mono<T> pollResult(Mono<Response<T>> responseMono, Class<T> responseBodyType) {
+        return responseMono.flatMap((Function<Response<T>, Mono<T>>) response -> {
+            if (response.getStatusCode() == 200) {
```
Comment (mkrueger) [inline sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java:321]: Are there constants for the status codes somewhere and if not wouldn't it make sense to introduce them?

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/WebAppsClientImpl.java [hunk 2]
```diff
@@ -370,7 +370,7 @@ Mono<Response<SiteConfigResourceCollection>> listConfigurations(@HostParam("$hos
 
         @Headers({ "Content-Type: application/json" })
         @Put("/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.Web/sites/{name}/config/appsettings")
-        @ExpectedResponses({ 200 })
+        @ExpectedResponses({ 200, 202 })
```
[the remark this answers is not in the document]
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/WebAppsClientImpl.java:373]: Yeah, it should be. I'll try that.

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/models/FunctionApp.java [hunk 3]
```diff
@@ -485,6 +510,52 @@ interface WithDockerContainerImage {
...
+             * Specifies the name of the <a href="https://learn.microsoft.com/en-us/azure/container-apps/environment">Azure Container Apps environment</a>
+             * to deploy the Function App on. The environment must be in the same resource group as the Function App.
+             * <p>When your container is hosted in a <a href="https://learn.microsoft.com/en-us/azure/container-apps/plans#consumption-dedicated">Consumption + Dedicated plan</a> structure,
+             * only the default Consumption plan is currently supported. Dedicated plans in this structure aren't yet supported for Functions.
+             * When running functions on Container Apps, you're charged only for the Container Apps usage.</p>
+             *
+             * @param managedEnvironmentName name of the Azure Container Apps environment
+             * @return the next stage of the definition
+             */
+            WithScaleRulesOrDockerContainerImage withManagedEnvironmentName(String managedEnvironmentName);
```
[the remark this answers is not in the document]
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/models/FunctionApp.java:537]: Discussed with Hanxiao, they can live without this method. 

Yeah, the implicit constraint is environment and function app be in the same resource group. I can't find it in docs, so probably drop this override for now.

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java (hunk 1 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java]: Switched to use directive and this method is no longer needed.

Yeah, it's a valid point. Will consider introducing if encountered next time, thanks!

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/WebAppsClientImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/WebAppsClientImpl.java:373]: directive PR: https://github.com/Azure/azure-rest-api-specs/pull/26935

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java [hunk 4]
```diff
@@ -171,13 +174,49 @@ public FunctionAppImpl withLatestRuntimeVersion() {
...
+                    addAppSettingIfNotModified(SETTING_WEB_JOBS_STORAGE, connectionString);
+                    addAppSettingIfNotModified(SETTING_WEB_JOBS_DASHBOARD, connectionString);
+                    return FunctionAppImpl.super.submitAppSettings();
+                }).then(
+                    Mono
+                        .fromCallable(
+                            () -> {
+                                resetStorageInfo();
+                                return this;
+                            }));
```
Comment (weidongxu-microsoft) [inline sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java:219]: Any difference from the code below? Can we extract them to a shared method?

**Review by weidongxu-microsoft — APPROVED** [review]
Approved to unblock.

Please check comment.

Also prefer to have tooling team take a look as well.

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java [hunk 5]
```diff
@@ -209,16 +248,23 @@ Mono<Indexable> submitAppSettings() {
...
+    private void resetStorageInfo() {
+        currentStorageAccount = storageAccountToSet;
+        storageAccountToSet = null;
+        storageAccountCreatable = null;
+    }
+
     @Override
     public OperatingSystem operatingSystem() {
+        if (isFunctionAppOnACA()) {
+            return OperatingSystem.LINUX;
```
[the remark this answers is not in the document]
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java:266]: Currently they only support linux. 

Sure, will add a comment, thanks!

File: sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java (hunk 4 above)
Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java:219]: For Function App on ACA, appServicePlanId() is null, resulting in NPE for this.manager().appServicePlans().getByIdAsync(this.appServicePlanId()). 
And two app settings `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING` and `WEBSITE_CONTENTSHARE` are not needed.

We should be able to combine them into one.

Comment (weidongxu-microsoft) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java:219]: Ok. If more difference, you can keep it, but better to add comment about the difference, so dev would know some configure is omitted in purpose.

Comment (XiaofeiCao) [reply sdk/resourcemanager/azure-resourcemanager-appservice/src/main/java/com/azure/resourcemanager/appservice/implementation/FunctionAppImpl.java]: Combined the two parts into one.

**Review by weidongxu-microsoft — APPROVED** [review]

**Review by Flanker32 — APPROVED** [review]
LGTM

**Pull request merged**